README.md
# flutter_window

一个可以拖拽缩放的窗口组件

## 📺  平台

| Android | Windows | Linux | Web |  MacOS | iOS |  
| --- | --- | --- | --- | --- | --- |
| ❌ | ❌ | ❌ | ✔️ | ❌ | ❌ |

> 时间有限，暂时只测试了Web平台

## 📦 安装
在`pubspec.yaml`文件添加依赖:

```yaml
dependencies:
  flutter_window: ^latest_version
```

## 💻 示例

```dart
import 'package:flutter/material.dart';
import 'package:flutter_window/property/border_property.dart';
import 'package:flutter_window/windows.dart';

void main() {
  runApp(const ToolmanApplication());
}

class ToolmanApplication extends StatefulWidget {
  const ToolmanApplication({Key? key}) : super(key: key);

  @override
  State<ToolmanApplication> createState() => _ToolmanApplicationState();
}

class _ToolmanApplicationState extends State<ToolmanApplication> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        brightness: Brightness.dark,
        primaryColor: Colors.blue,
      ),
      home: Scaffold(
        body: LayoutBuilder(
          builder: (context, constraints) {
            return SizedBox(
              width: constraints.maxWidth,
              height: constraints.maxHeight,
              // color: Colors.black,
              child: Stack(
                children: const [
                  Windows(
                    draggable: false,
                    border: BorderProperty.onlyThickness(3),
                    height: 500,
                    width: 500,
                    body: Text('Emulates the operating system window,'
                        ' which can be zoomed and dragged'),
                    index: 1,
                    position: Offset(50, 50),
                  )
                ],
              ),
            );
          },
        ),
      ),
    );
  }
}
```
> 组件必须用 `Stack` 组件包裹，不然无法拖动，更多属性参考源码。

## ⚙️属性

| 属性值 | 类型 |说明                           |
| ------| ---- |--------------------------------- |
| index | int | 窗口组件索引（预留）|
| header | widget? | 自定义头部，不设置默认包含全屏按钮和标题 |
| body | widget | 窗口主体内容 |
| title | string? | 窗口标题 |
| height | double? | 窗口高度，默认值 350 |
| width | double? | 窗口宽度， 默认值 350 |
| draggable | bool? | 是否可拖动 |
| isFullscreen | bool? | 是否全屏 |
| border | BorderProperty? | 具体属性查看`BorderProperty` |
| position | Offset? | 窗口偏移值 |


## 📢 最后
这算是我的一个Flutter练手代码，刚接触Flutter不久还有很多不熟悉，如果有哪些地方写的不好，还望各位大佬包涵。
也欢迎各路大神指点迷津。

## 开源协议
本项目基于**BSD3**协议

CHANGELOG.md
## 1.0.0
* 基础版本窗口组件

lib/windows.dart
import 'package:flutter/material.dart';
import 'package:flutter_window/border/windows_border_box.dart';
import 'package:flutter_window/property/border_property.dart';
import 'package:flutter_window/windows_state.dart';

class Windows extends StatefulWidget {
  const Windows({
    Key? key,
    required this.index,
    required this.body,
    this.title = "Flutter Toolman Windows",
    this.height = 350,
    this.width = 350,
    this.draggable = true,
    this.isFullscreen = false,
    this.header,
    this.border = const BorderProperty(
      color: BorderColor.def(),
      borderThickness: 1,
    ),
    this.position = const Offset(0, 0),
  }) : super(key: key);

  final String? title;

  final double? height;

  final double? width;

  /// 自定义头部组件，不存在则默认
  final Widget? header;

  final Widget body;

  /// 是否可拖动，悬浮状态可以拖动
  final bool? draggable;

  /// 是否全屏
  final bool? isFullscreen;

  /// 窗口位置
  final Offset? position;

  /// 边框属性
  final BorderProperty? border;

  /// 层级，数字越大则在最顶层
  final int index;

  @override
  State<Windows> createState() => _WindowsState();
}

class _WindowsState extends State<Windows> with SingleTickerProviderStateMixin {
  final GlobalKey windowsKey = GlobalKey();
  late final WindowsState _oldState = WindowsState();
  late Offset _position;
  final double _minHeight = 350;
  final double _minWidth = 350;

  final double _headerHeight = 35;

  late bool _isFullscreen;

  late double _windowsHeight;
  late double _windowsWidth;

  /// 边框厚度
  late double _borderThickness;
  final Color _borderColor = const Color(0xFF595959);

  @override
  void initState() {
    super.initState();

    _borderThickness = widget.border!.borderThickness;
    _isFullscreen = widget.isFullscreen!;
    _windowsHeight = widget.height! < _minWidth ? _minWidth : widget.height!;
    _windowsWidth = widget.width! < _minWidth ? _minWidth : widget.width!;
    _position = widget.position!;

    _setOldState();
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    _checkCritical();
    _setOldState();
  }

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return _buildWindows(context);
  }

  /// 检查临界值，位置和宽高是否超出窗口
  /// 超出则挪到可视位置
  void _checkCritical() {
    Size controllerSize = MediaQuery.of(context).size;
    var criticalPointX = _position.dx + _windowsWidth;
    var criticalPointY = _position.dy + _windowsHeight;
    var newDx = _position.dx;
    var newDy = _position.dy;
    if (criticalPointX > controllerSize.width) {
      newDx = criticalPointX - controllerSize.width;
    }
    if (criticalPointY > controllerSize.height) {
      newDy = criticalPointY - controllerSize.height;
    }
    _position = Offset(newDx, newDy);
  }

  /// 保存窗口前一次变化的属性
  void _setOldState() {
    _oldState
      ..windowsHeight = _windowsHeight
      ..windowsWidth = _windowsWidth
      ..dy = _position.dy
      ..dx = _position.dx;
  }

  Widget _buildWindows(BuildContext context) {
    Size controllerSize = MediaQuery.of(context).size;
    return Positioned(
      key: windowsKey,
      left: _position.dx,
      top: _position.dy,
      child: ConstrainedBox(
        constraints: BoxConstraints(
          maxHeight: controllerSize.height,
          maxWidth: controllerSize.width,
          minHeight: _minHeight,
          minWidth: _minWidth,
        ),
        child: SizedBox(
          height: _windowsHeight,
          width: _windowsWidth,
          child: Stack(
            children: [
              WindowsBorderBox(
                parentKey: windowsKey,
                windowsHeight: _windowsHeight,
                windowsWidth: _windowsWidth,
                minHeight: _minHeight,
                minWidth: _minWidth,
                borderProperty: widget.border!,
                overlayState: Overlay.of(
                  context,
                  debugRequiredFor: widget,
                  rootOverlay: false,
                )!,
                onEnd: (width, height, offset) {
                  _windowsWidth = width;
                  _windowsHeight = height;
                  _position = offset;
                  _setOldState();
                  setState(() {});
                },
              ),
              Padding(
                padding: EdgeInsets.all(_borderThickness),
                child: Column(
                  children: [
                    _buildHeader(),
                    _buildBody(),
                  ],
                ),
              )
            ],
          ),
        ),
      ),
    );
  }

  /// 创建顶部工具栏
  Widget _buildHeader() {
    return Material(
      child: Draggable(
        onDraggableCanceled: (Velocity velocity, Offset offset) {
          if (!widget.draggable!) {
            return;
          }
          _position = offset;
          _setOldState();

          setState(() {});
        },
        feedback: _buildDragFeedback(_windowsHeight, _windowsWidth),
        child: Container(
          width: _windowsWidth - _borderThickness * 2,
          height: _headerHeight,
          color: Colors.transparent,
          alignment: Alignment.centerLeft,
          child: widget.header ??
              Row(
                mainAxisAlignment: MainAxisAlignment.end,
                children: [
                  Padding(
                    padding: const EdgeInsets.fromLTRB(12, 0, 12, 0),
                    child: Text("${widget.title}"),
                  ),
                  _buttonFullscreen(),
                ],
              ),
        ),
      ),
    );
  }

  /// 创建显示内容区域
  Widget _buildBody() {
    return Expanded(
      child: Container(
        color: const Color(0xFF434343),
        width: _windowsWidth - _borderThickness * 2,
        height: _windowsHeight - _headerHeight,
        child: Padding(
          padding: const EdgeInsets.all(12),
          child: widget.body,
        ),
      ),
    );
  }

  Widget _buildDragFeedback(double height, double width) {
    return ConstrainedBox(
      constraints: BoxConstraints(
        maxHeight: height,
        maxWidth: width,
      ),
      child: Container(
        decoration: BoxDecoration(
          color: _borderColor.withOpacity(0.3),
          border: Border.all(color: _borderColor, style: BorderStyle.solid),
        ),
      ),
    );
  }

  /// ********************************
  /// 铺满屏幕按钮
  Widget _buttonFullscreen() {
    return IconButton(
      onPressed: () {
        if (_isFullscreen) {
          _windowsWidth = _oldState.windowsWidth;
          _windowsHeight = _oldState.windowsHeight;
          _position = Offset(_oldState.dx, _oldState.dy);
        } else {
          final size = MediaQuery.of(context).size;
          _windowsWidth = size.width;
          _windowsHeight = size.height;
          _position = const Offset(0, 0);
        }

        _isFullscreen = !_isFullscreen;

        setState(() {});
      },
      icon: Icon(
        !_isFullscreen ? Icons.fullscreen : Icons.fullscreen_exit,
        size: 20,
      ),
    );
  }
}


lib/property/border_property.dart
import 'package:flutter/material.dart';

/// 边框的参数属性值
class BorderProperty {
  const BorderProperty({
    required this.color,
    required this.borderThickness,
  });

  /// 只设置边框的厚度，颜色取默认值
  const BorderProperty.onlyThickness(this.borderThickness) : color = const BorderColor.def();

  /// 边框颜色
  final BorderColor color;

  /// 边框厚度，也用作内容的内边距以免被遮挡
  final double borderThickness;
}

class BorderColor {
  final Color leftColor;
  final Color topColor;
  final Color rightColor;
  final Color bottomColor;

  static const Color defaultColor = Color(0xFF595959);

  const BorderColor.def()
      : leftColor = defaultColor,
        topColor = defaultColor,
        rightColor = defaultColor,
        bottomColor = defaultColor;

  const BorderColor.none()
      : leftColor = Colors.transparent,
        topColor = Colors.transparent,
        rightColor = Colors.transparent,
        bottomColor = Colors.transparent;

  const BorderColor.all(Color color)
      : leftColor = color,
        topColor = color,
        rightColor = color,
        bottomColor = color;
}


lib/border/windows_border_box.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_window/border/windows_border_item.dart';
import 'package:flutter_window/direction.dart';
import 'package:flutter_window/property/border_property.dart';

typedef BorderClickCallback = void Function(Offset offset, Direction direction);
typedef BorderDragCallback = void Function(Offset offset, Direction direction);
typedef BorderEndCallback = void Function(
    double width, double height, Offset offset);

class WindowsBorderBox extends StatefulWidget {
  const WindowsBorderBox({
    Key? key,
    required this.windowsWidth,
    required this.windowsHeight,
    required this.minWidth,
    required this.minHeight,
    required this.parentKey,
    required this.overlayState,
    required this.borderProperty,
    this.onClick,
    this.onDrag,
    this.onEnd,
  }) : super(key: key);

  final GlobalKey parentKey;

  final double windowsWidth;

  final double windowsHeight;

  final double minWidth;

  final double minHeight;

  final OverlayState overlayState;

  final BorderProperty borderProperty;

  final BorderClickCallback? onClick;

  final BorderDragCallback? onDrag;

  final BorderEndCallback? onEnd;

  @override
  State<WindowsBorderBox> createState() => _WindowsBorderBoxState();
}

class _WindowsBorderBoxState extends State<WindowsBorderBox> {
  /// 需要计算的最终偏移值
  late Offset _feedbackOffset;

  /// 点击边框时候，保存当前的光标样式
  late SystemMouseCursor _resizeCursor;

  /// 触发宽高重置的边框线长度
  final double _cornerWidth = 20;

  /// 边框厚度
  late double _borderThickness;

  /// 边框颜色样式
  late BorderColor _borderColor;

  /// 点击边框记录当前坐标，用于计算偏移值
  late Offset _startPosition = Offset.zero;

  /// 限制偏移量，缩小不能少于这个参数值
  late Offset _shrinkLimitOffset;

  OverlayEntry? _entry;

  late double _windowsWidth;

  late double _windowsHeight;

  @override
  void initState() {
    super.initState();

    _windowsWidth = widget.windowsWidth;
    _windowsHeight = widget.windowsHeight;
    _borderColor = widget.borderProperty.color;
    _borderThickness = widget.borderProperty.borderThickness;
    _feedbackOffset = Offset.zero;
  }

  @override
  Widget build(BuildContext context) {
    return _buildWindowsBorder();
  }

  @override
  void dispose() {
    super.dispose();
    _clearPoint();
  }

  void _onClick(Offset position, Direction direction, cursor) {
    _buildPoint();
    _startPosition = position;
    _resizeCursor = cursor;
    _shrinkLimitOffset = Offset(
      widget.windowsWidth - widget.minWidth,
      widget.windowsHeight - widget.minHeight,
    );
  }

  void _onMoveStart(Offset position, Direction direction) {
    handleBorderMove(position, direction);
  }

  void _onMoveEnd() {
    widget.onEnd!(
      _windowsWidth,
      _windowsHeight,
      _getFinalOffset(_feedbackOffset),
    );
    _clearPoint();
  }

  Widget _buildWindowsBorder() {
    return Container(
      color: Colors.transparent,
      child: Stack(
        children: [
          Positioned(
            top: 0,
            left: 0,
            child: _buildHorizontalBorder(Direction.top),
          ),
          Positioned(
            right: 0,
            top: 0,
            child: _buildVerticalBorder(Direction.right),
          ),
          Positioned(
            bottom: 0,
            left: 0,
            child: _buildHorizontalBorder(Direction.bottom),
          ),
          Positioned(
            left: 0,
            top: 0,
            child: _buildVerticalBorder(Direction.left),
          ),
        ],
      ),
    );
  }

  /// 创建水平边框
  Widget _buildHorizontalBorder(Direction direction) {
    return Row(
      children: [
        WindowsBorderItem(
          cursor: Direction.top == direction
              ? SystemMouseCursors.resizeUpLeft
              : SystemMouseCursors.resizeDownLeft,
          height: _borderThickness,
          width: _cornerWidth,
          direction: Direction.top == direction
              ? Direction.topLeft
              : Direction.bottomLeft,
          color: _borderColor.topColor,
          onClick: _onClick,
          onMoveStart: _onMoveStart,
          onMoveEnd: _onMoveEnd,
        ),
        WindowsBorderItem(
          cursor: Direction.top == direction
              ? SystemMouseCursors.resizeUp
              : SystemMouseCursors.resizeDown,
          height: _borderThickness,
          width: widget.windowsWidth - _cornerWidth * 2,
          direction: direction,
          color: _borderColor.topColor,
          onClick: _onClick,
          onMoveStart: _onMoveStart,
          onMoveEnd: _onMoveEnd,
        ),
        WindowsBorderItem(
          cursor: Direction.top == direction
              ? SystemMouseCursors.resizeUpRight
              : SystemMouseCursors.resizeDownRight,
          height: _borderThickness,
          width: _cornerWidth,
          direction: Direction.top == direction
              ? Direction.topRight
              : Direction.bottomRight,
          color: _borderColor.topColor,
          onClick: _onClick,
          onMoveStart: _onMoveStart,
          onMoveEnd: _onMoveEnd,
        ),
      ],
    );
  }

  /// 创建垂直边框
  Widget _buildVerticalBorder(Direction direction) {
    return Column(
      children: [
        WindowsBorderItem(
          cursor: Direction.left == direction
              ? SystemMouseCursors.resizeUpLeft
              : SystemMouseCursors.resizeUpRight,
          height: _cornerWidth,
          width: _borderThickness,
          direction: Direction.left == direction
              ? Direction.leftTop
              : Direction.rightTop,
          color: _borderColor.topColor,
          onClick: _onClick,
          onMoveStart: _onMoveStart,
          onMoveEnd: _onMoveEnd,
        ),
        WindowsBorderItem(
          cursor: Direction.left == direction
              ? SystemMouseCursors.resizeLeft
              : SystemMouseCursors.resizeRight,
          height: widget.windowsHeight - _cornerWidth * 2,
          width: _borderThickness,
          direction: direction,
          color: _borderColor.topColor,
          onClick: _onClick,
          onMoveStart: _onMoveStart,
          onMoveEnd: _onMoveEnd,
        ),
        WindowsBorderItem(
          cursor: Direction.left == direction
              ? SystemMouseCursors.resizeDownLeft
              : SystemMouseCursors.resizeDownRight,
          height: _cornerWidth,
          width: _borderThickness,
          direction: Direction.left == direction
              ? Direction.leftBottom
              : Direction.rightBottom,
          color: _borderColor.topColor,
          onClick: _onClick,
          onMoveStart: _onMoveStart,
          onMoveEnd: _onMoveEnd,
        ),
      ],
    );
  }

  void handleBorderMove(Offset position, Direction direction) {
    switch (direction) {
      case Direction.top:
        _calculate(position, 0, -1, true, false);
        break;
      case Direction.topLeft:
        _calculate(position, -1, -1, false, false);
        break;
      case Direction.topRight:
        _calculate(position, 1, -1, true, false);
        break;
      case Direction.bottom:
        _calculate(position, 0, 1, true, true);
        break;
      case Direction.bottomLeft:
        _calculate(position, -1, 1, false, true);
        break;
      case Direction.bottomRight:
        _calculate(position, 1, 1, true, true);
        break;
      case Direction.left:
        _calculate(position, -1, 0, false, true);
        break;
      case Direction.leftTop:
        _calculate(position, -1, -1, false, false);
        break;
      case Direction.leftBottom:
        _calculate(position, -1, 1, false, true);
        break;
      case Direction.right:
        _calculate(position, 1, 0, true, true);
        break;
      case Direction.rightTop:
        _calculate(position, 1, -1, true, false);
        break;
      case Direction.rightBottom:
        _calculate(position, 1, 1, true, true);
        break;
      default:
        break;
    }
  }

  /// 计算距离和偏移量
  /// factor是XY轴的计算系数，[1,-1,0]
  void _calculate(
      Offset position, int factorX, int factorY, bool fixedX, bool fixedY) {
    double distanceX = 0.0;
    double distanceY = 0.0;

    distanceX = position.dx - _startPosition.dx;
    distanceY = position.dy - _startPosition.dy;

    _feedbackOffset = Offset(
        fixedX ? 0 : min(distanceX, _shrinkLimitOffset.dx) * factorX * -1,
        fixedY ? 0 : min(distanceY, _shrinkLimitOffset.dy) * factorY * -1);

    _windowsHeight =
        max(widget.windowsHeight + distanceY * factorY, widget.minHeight);
    _windowsWidth =
        max(widget.windowsWidth + distanceX * factorX, widget.minWidth);

    _updateParentState();
  }

  Offset _getFinalOffset(Offset position) {
    var controllerSize = MediaQuery.of(context).size;
    var parentLocal = _getParentPosition();

    var finalDx = parentLocal.dx + position.dx;
    var finalDy = parentLocal.dy + position.dy;

    if (finalDx <= 0) {
      finalDx = 0;
    } else if (finalDx + _windowsWidth >= controllerSize.width) {
      finalDx = controllerSize.width - _windowsWidth;
    }

    if (finalDy <= 0) {
      finalDy = 0;
    } else if (finalDy + _windowsHeight >= controllerSize.height) {
      finalDy = controllerSize.height - _windowsHeight;
    }

    return Offset(finalDx, finalDy);
  }

  void _updateParentState() {
    widget.overlayState.setState(() {});
  }

  void _clearPoint() {
    _entry?.remove();
    _entry = null;
    _clearResize();
  }

  void _buildPoint() {
    _clearPoint();
    var parentLocal = _getParentPosition();

    _entry = OverlayEntry(
      builder: (context) {
        return Positioned(
          left: parentLocal.dx + _feedbackOffset.dx,
          top: parentLocal.dy + _feedbackOffset.dy,
          child: SizedBox(
            height: _windowsHeight,
            width: _windowsWidth,
            child: MouseRegion(
              cursor: _resizeCursor,
              child: Container(
                decoration: BoxDecoration(
                  color: Colors.grey.withOpacity(0.3),
                ),
              ),
            ),
          ),
        );
      },
    );

    widget.overlayState.insert(_entry!);
  }

  void _clearResize() {
    _feedbackOffset = Offset.zero;
    _windowsWidth = widget.windowsWidth;
    _windowsHeight = widget.windowsHeight;
  }

  RenderBox _getParentBox() {
    return widget.parentKey.currentContext?.findRenderObject() as RenderBox;
  }

  Offset _getParentPosition() {
    return _getParentBox().localToGlobal(Offset.zero);
  }
}


lib/border/windows_border_item.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_window/direction.dart';

typedef BorderItemClickCallback = void Function(
    Offset position, Direction direction, SystemMouseCursor cursor);

typedef BorderItemMoveStartCallback = void Function(
    Offset position, Direction direction);

typedef BorderItemMoveEndOverCallback = void Function();

class WindowsBorderItem extends StatefulWidget {
  const WindowsBorderItem({
    Key? key,
    required this.direction,
    required this.width,
    required this.height,
    this.onClick,
    this.onMoveStart,
    this.onMoveEnd,
    this.cursor = SystemMouseCursors.click,
    this.color = const Color(0xFF595959),
  }) : super(key: key);

  final Direction direction;

  final double width;

  final double height;

  final BorderItemClickCallback? onClick;

  final BorderItemMoveStartCallback? onMoveStart;

  final BorderItemMoveEndOverCallback? onMoveEnd;

  final SystemMouseCursor cursor;

  final Color color;

  @override
  State<WindowsBorderItem> createState() => _WindowsBorderItemState();
}

class _WindowsBorderItemState extends State<WindowsBorderItem> {
  @override
  void initState() {
    super.initState();
  }

  @override
  void dispose() {
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MouseRegion(
      cursor: widget.cursor,
      child: GestureDetector(
        onPanDown: (event) {
          widget.onClick!(event.globalPosition, widget.direction, widget.cursor);
        },
        onPanUpdate: (event) {
          widget.onMoveStart!(event.globalPosition, widget.direction);
        },
        onPanCancel: () {
          widget.onMoveEnd!();
        },
        onPanEnd: (event) {
          widget.onMoveEnd!();
        },
        child: Container(
            color: widget.color, width: widget.width, height: widget.height),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_window/property/border_property.dart';
import 'package:flutter_window/windows.dart';

void main() {
  runApp(const ToolmanApplication());
}

class ToolmanApplication extends StatefulWidget {
  const ToolmanApplication({Key? key}) : super(key: key);

  @override
  State<ToolmanApplication> createState() => _ToolmanApplicationState();
}

class _ToolmanApplicationState extends State<ToolmanApplication> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        brightness: Brightness.dark,
        primaryColor: Colors.blue,
      ),
      home: Scaffold(
        body: LayoutBuilder(
          builder: (context, constraints) {
            return SizedBox(
              width: constraints.maxWidth,
              height: constraints.maxHeight,
              // color: Colors.black,
              child: Stack(
                children: const [
                  Windows(
                    draggable: false,
                    border: BorderProperty.onlyThickness(3),
                    height: 500,
                    width: 500,
                    body: Text('Emulates the operating system window,'
                        ' which can be zoomed and dragged'),
                    index: 1,
                    position: Offset(50, 50),
                  )
                ],
              ),
            );
          },
        ),
      ),
    );
  }
}
